### PR TITLE
Types/crunch structs

### PR DESCRIPTION
**Pull request by mejango**
# Description

Side quest to optimize struct types and clean. I've kept all tests passing along the way.

JBController ran into a contract size constraint, hence the few refactors into internal view fns.

## Limitations & risks

*Are there any trade-off or new vulnarbility surface based on theses changes?*

# Check-list
- [ ] Tests are covering the new feature
- [ ] Code is [natspec'd](https://docs.soliditylang.org/en/v0.8.17/natspec-format.html)
- [ ] Code is [linted and formatted](https://docs.soliditylang.org/en/v0.8.17/style-guide.html)
- [ ] I have run the test locally (and they pass)
- [ ] I have rebased to the latest main commit (and tests still pass)

# Interactions
These changes will impact the following contracts:
- Directly:

- Indirectly:

### CODE REVIEW

File: src/JBFundAccessLimits.sol [hunk 1]
```diff
@@ -338,7 +338,7 @@ contract JBFundAccessLimits is JBControlled, IJBFundAccessLimits {
                 // Set the surplus allowance if there is one.
                 if (surplusAllowance.amount > 0) {
                     _packedSurplusAllowancesDataOf[projectId][rulesetId][fundAccessLimitGroup[i].terminal][fundAccessLimitGroup[i]
-                        .token].push(surplusAllowance.amount | (surplusAllowance.currency << 224));
+                        .token].push(uint256(surplusAllowance.amount) | (uint256(surplusAllowance.currency) << 224));
```
Comment (simplemachine92) [inline src/JBFundAccessLimits.sol:341]: Since the amount & currency packing is performed the same for surplus and payouts, we could declare this under an internal utility function. But also contract size isnt a problem here.

File: src/JBPermissions.sol [hunk 2]
```diff
@@ -224,16 +224,14 @@ contract JBPermissions is IJBPermissions {
     /// @notice Converts an array of permission IDs to a packed `uint256`.
     /// @param permissionIds The IDs of the permissions to pack.
     /// @return packed The packed value.
-    function _packedPermissions(uint256[] calldata permissionIds) internal pure returns (uint256 packed) {
+    function _packedPermissions(uint8[] calldata permissionIds) internal pure returns (uint256 packed) {
         // Keep a reference to the permission being iterated on.
         uint256 permissionId;
```
Comment (simplemachine92) [inline src/JBPermissions.sol:229]: could reduce the size in memory here?

File: src/JBRulesets.sol [hunk 3]
```diff
@@ -952,27 +953,27 @@ contract JBRulesets is JBControlled, IJBRulesets {
...
-        // `basedOnId` in bits 88-143 bits.
-        ruleset.basedOnId = uint256(uint56(packedIntrinsicProperties >> 88));
-        // `start` in bits 144-199 bits.
-        ruleset.start = uint256(uint56(packedIntrinsicProperties >> 144));
-        // `cycleNumber` in bits 200-255 bits.
-        ruleset.cycleNumber = uint256(uint56(packedIntrinsicProperties >> 200));
+        // `weight` in bits 0-111 bits.
+        ruleset.weight = uint112(packedIntrinsicProperties);
+        // `basedOnId` in bits 112-159 bits.
+        ruleset.basedOnId = uint48(packedIntrinsicProperties >> 112);
```
Comment (simplemachine92) [inline src/JBRulesets.sol:963]: Should we have a lib for all of these pointer indexes?

File: src/JBSplits.sol [hunk 4]
```diff
@@ -273,9 +265,9 @@ contract JBSplits is JBControlled, IJBSplits {
             // `preferAddToBalance` in bit 0.
             split.preferAddToBalance = packedSplitPart1 & 1 == 1;
             // `percent` in bits 1-32.
-            split.percent = uint256(uint32(packedSplitPart1 >> 1));
+            split.percent = uint32(packedSplitPart1 >> 1);
             // `projectId` in bits 33-88.
-            split.projectId = uint256(uint56(packedSplitPart1 >> 33));
+            split.projectId = uint56(packedSplitPart1 >> 33);
```
Comment (simplemachine92) [inline src/JBSplits.sol:270]: Again, lib for pointer indexes?

File: src/JBSplits.sol [hunk 5]
```diff
@@ -172,9 +170,6 @@ contract JBSplits is JBControlled, IJBSplits {
             // The percent should be greater than 0.
             if (splits[i].percent == 0) revert INVALID_SPLIT_PERCENT();
 
-            // `projectId` should fit within a uint56
```
Comment (simplemachine92) [inline src/JBSplits.sol:175]: Why is this removed? setSplitGroupsOf in controller still accepts a uint256 projectId

File: test/units/static/JBSplits/TestSetSplitGroupsOf.sol [hunk 6]
```diff
@@ -116,31 +116,6 @@ contract TestSetSplitGroupsOf_Local is JBSplitsSetup {
         _splits.setSplitGroupsOf(_projectId, _rulesetId, _splitsGroup);
     }
 
-    function test_GivenProjectIdGtUint56Max() external whenCallerIsController {
```
Comment (simplemachine92) [inline test/units/static/JBSplits/TestSetSplitGroupsOf.sol:119]: See my above comment about setSplitGroupsOf

**Review by simplemachine92 — COMMENTED** [review]
Generally a lot cleaner but some functions still accept values that don't respect some struct assumptions. i.e. projectId in setSplitGroupsOf from JBC -> JBSplits could have a projectId of up to uint256.max.

That's worrisome as it looks like some important sanity checks could have been lost.

File: src/JBController.sol [hunk 7]
```diff
@@ -444,32 +485,24 @@ contract JBController is JBPermissioned, ERC2771Context, IJBController, IJBMigra
...
             account: PROJECTS.ownerOf(projectId),
             projectId: projectId,
             permissionId: JBPermissionIds.MINT_TOKENS,
-            alsoGrantAccessIf: DIRECTORY.isTerminalOf(projectId, IJBTerminal(_msgSender()))
-                || _msgSender() == ruleset.dataHook()
-                || (
-                    ruleset.dataHook() != address(0)
-                        && IJBRulesetDataHook(ruleset.dataHook()).hasMintPermissionFor(projectId, _msgSender())
-                )
+            alsoGrantAccessIf: _isTerminalOf(projectId, _msgSender()) || _msgSender() == ruleset.dataHook()
```
[the remark this answers is not in the document]
Comment (mejango) [reply src/JBController.sol:496]: did the logic change?

grant access if:
A: the terminal is calling
B: or, the data hook is called
C: or, there's a data hook and it explicitly gives permissions.

seems right?

File: src/JBSplits.sol (hunk 5 above)
Comment (mejango) [reply src/JBSplits.sol:175]: the splits.projectId is guaranteed within a uint56 because the type is such. This comment and the line underneath are not necessary.

File: src/JBRulesets.sol (hunk 3 above)
Comment (mejango) [reply src/JBRulesets.sol:963]: ah, the indexes aren't relevant to the outside world, it's just packing details.

File: src/JBFundAccessLimits.sol (hunk 1 above)
Comment (mejango) [reply src/JBFundAccessLimits.sol:341]: true. im cool with it as is.

File: src/JBSplits.sol (hunk 5 above)
Comment (mejango) [reply src/JBSplits.sol:175]: image

the projectId that is 256 is the one that stores the mapping. the projectId in the split, which is being checked here, is uint56.

File: src/JBMultiTerminal.sol [hunk 8]
```diff
@@ -1486,12 +1486,15 @@ contract JBMultiTerminal is JBPermissioned, ERC2771Context, IJBMultiTerminal {
...
 
+        // Keep a reference to the fee for the specification's amount.
+        uint256 specificationAmountFee;
+
         for (uint256 i; i < numberOfSpecifications; i++) {
             // Set the specification being iterated on.
             specification = specifications[i];
 
             // Get the fee for the specified amount.
-            uint256 specificationAmountFee =
```
Comment (xBA5ED) [inline src/JBMultiTerminal.sol:1494]: nit: I think I like the old style here better, just because it makes it harder to make a mistake in the future and we don't need this value outside of the scope of the loop.

File: src/JBRulesets.sol [hunk 9]
```diff
@@ -388,17 +388,17 @@ contract JBRulesets is JBControlled, IJBRulesets {
             revert INVALID_DECAY_RATE();
         }
 
-        // Weight must fit into a uint88.
-        if (weight > type(uint88).max) revert INVALID_WEIGHT();
+        // Weight must fit into a uint112.
+        if (weight > type(uint112).max) revert INVALID_WEIGHT();
```
Comment (xBA5ED) [inline src/JBRulesets.sol:392]: Should we make an `JBConstants.MAX_WEIGHT`?

**Review by xBA5ED — APPROVED** [review]

File: src/JBMultiTerminal.sol (hunk 8 above)
Comment (mejango) [reply src/JBMultiTerminal.sol:1494]: got it. will change

File: src/JBRulesets.sol (hunk 9 above)
Comment (mejango) [reply src/JBRulesets.sol:392]: meh. not really useful for any division calculations

File: src/JBPermissions.sol (hunk 2 above)
Comment (mejango) [reply src/JBPermissions.sol:229]: not sure here. i reduced it to make it easier to pass data around from the struct

File: test/units/static/JBSplits/TestSetSplitGroupsOf.sol (hunk 6 above)
Comment (mejango) [reply test/units/static/JBSplits/TestSetSplitGroupsOf.sol:119]: settled in that comment
